compiler/src/main/kotlin/motif/compiler/codegenv2/KotlinCodeGenerator.kt

```diff
@@ -0,0 +1,282 @@
+                .add("%N = %L", cacheFieldName, instantiation.spec())
+                .endControlFlow()
+                .endControlFlow()
+                .endControlFlow()
+                .add("return %N as %T", cacheFieldName, returnTypeName.kt)
+                .build()
+    }
+
+    private fun FactoryProviderMethodBody.Uncached.spec(): CodeBlock {
+        return CodeBlock.of("return %L", instantiation.spec())
+    }
+
+    private fun FactoryProviderInstantiation.spec(): CodeBlock {
+        return when (this) {
+            is FactoryProviderInstantiation.Basic -> spec()
+            is FactoryProviderInstantiation.Constructor -> spec()
+            is FactoryProviderInstantiation.Binds -> spec()
+        }
+    }
+
+    private fun FactoryProviderInstantiation.Basic.spec(): CodeBlock {
+        return if (isStatic) {
+            CodeBlock.of("%T.%N%L", objectsClassName.kt, factoryMethodName, callProviders.spec())
+        } else {
+            CodeBlock.of("%N.%N%L", objectsFieldName, factoryMethodName, callProviders.spec())
+        }
+    }
+
+    private fun FactoryProviderInstantiation.Constructor.spec(): CodeBlock {
+        return CodeBlock.of("%T%L", returnTypeName.kt, callProviders.spec())
+    }
+
+    private fun FactoryProviderInstantiation.Binds.spec(): CodeBlock {
+        return CodeBlock.of("%N()", providerMethodName)
+    }
+
+    private fun CallProviders.spec(): String {
+        val callString = providerMethodNames.joinToString { "$it()" }
+        return "($callString)"
+    }
+
+    private fun SpreadProviderMethod.spec(): FunSpec {
+        return FunSpec.builder(name).apply {
+            returns(returnTypeName.kt)
+            if (isStatic) {
+                addStatement("return %T.%N()", sourceTypeName.kt, spreadMethodName)
+            } else {
+                addStatement("return %N().%N()", sourceProviderMethodName, spreadMethodName)
+            }
+        }.build()
+    }
+
+    private fun DependencyProviderMethod.spec(): FunSpec {
+        return FunSpec.builder(name)
+                .returns(returnTypeName.kt)
+                .addStatement("return %N.%N()", dependenciesFieldName, dependencyMethodName)
+                .build()
+    }
+
+    private fun Dependencies.spec(): TypeSpec {
+        return TypeSpec.interfaceBuilder(className.kt).apply {
+            addModifiers(KModifier.PUBLIC)
+            methods.forEach { addFunction(it.spec()) }
+        }.build()
+    }
+
+    private fun DependencyMethod.spec(): FunSpec {
+        return FunSpec.builder(name).apply {
+            qualifier?.let { addAnnotation(it.spec()) }
+            addModifiers(KModifier.PUBLIC, KModifier.ABSTRACT)
+            returns(returnTypeName.kt)
+            addKdoc(javaDoc.spec())
+        }.build()
+    }
+
+    private fun Qualifier.spec(): AnnotationSpec {
+        return AnnotationSpec.get(annotation.mirror)
+    }
+
+    private fun DependencyMethodJavaDoc.spec(): CodeBlock {
```

Comment:
Should we generate KDoc here instead of JavaDoc?

Reply:
Ah right - I updated the link format but I suppose kdoc supports simple markdown lists?

Reply:
Yup. I believe so.

Reply:
Updated